File: sdk/javascript/src/adapter.ts
// Adapter (LoRA) wrapper — wraps pie:core/adapter WIT resource.

import { Adapter as _Adapter } from 'pie:core/adapter';

/**
 * A LoRA adapter instance.
 *
 * Wraps the `pie:core/adapter.Adapter` WIT resource. Implements
 * `Disposable` for use with `using`:
 *
 *     using adapter = Adapter.create("my-lora");
 *     adapter.load("/path/to/weights");
 *     // adapter.destroy() called automatically on scope exit
 */
export class Adapter implements Disposable {
    /** @internal */
    readonly _handle: _Adapter;

    private constructor(handle: _Adapter) {
        this._handle = handle;
    }

    /** Disposable protocol — calls `destroy()`. */
    [Symbol.dispose](): void {
        this.destroy();
    }

    /** Create a new adapter for the model with the given name. */
    static create(name: string): Adapter {
        return new Adapter(_Adapter.create(name));
    }

    /** Open an existing adapter by name. Returns `undefined` if not found. */
    static open(name: string): Adapter | undefined {
        const handle = _Adapter.open(name);
        return handle !== undefined ? new Adapter(handle) : undefined;
    }

    /** Destroy the adapter, releasing its resources. */
    destroy(): void {
        this._handle.destroy();
    }

    /** Fork this adapter with a new name. */
    fork(name: string): Adapter {
        return new Adapter(this._handle.fork(name));
    }

    /** Load adapter weights from a file path. */
    load(path: string): void {
        this._handle.load(path);
    }

    /** Save adapter weights to a file path. */
    save(path: string): void {
        this._handle.save(path);
    }
}

File: sdk/python/src/inferlet/adapter.py
"""
Adapter resource wrapper for ``pie:core/adapter``.

Supports loading, saving, cloning, and locking model adapters (e.g. LoRA).
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from wit_world.imports import adapter as _adapter

if TYPE_CHECKING:
    from .model import Model


class Adapter:
    """Wraps the WIT adapter resource.

    Usage::

        adapter = Adapter.create(model, "my-adapter")
        adapter.load("/path/to/weights")
        adapter.save("/path/to/output")
    """

    __slots__ = ("_handle",)

    def __init__(self, handle: _adapter.Adapter) -> None:
        self._handle = handle

    @staticmethod
    def create(model: Model, name: str) -> Adapter:
        """Create a new adapter for a model."""
        return Adapter(_adapter.Adapter.create(model._handle, name))

    @staticmethod
    def open(model: Model, name: str) -> Adapter | None:
        """Open an existing adapter by name."""
        raw = _adapter.Adapter.open(model._handle, name)
        if raw is None:
            return None
        return Adapter(raw)

    def fork(self, new_name: str) -> Adapter:
        """Fork this adapter with a new name."""
        return Adapter(self._handle.fork(new_name))

    def load(self, path: str) -> None:
        """Load adapter weights from a file path."""
        self._handle.load(path)

    def save(self, path: str) -> None:
        """Save adapter weights to a file path."""
        self._handle.save(path)

    def __enter__(self) -> Adapter:
        return self

    def __exit__(self, *args) -> None:
        pass

    def __repr__(self) -> str:
        return f"Adapter({id(self._handle):#x})"
